File: binary_tree/levelOrder.cpp
// leetcode 102. levelOrder bfs cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) 
    {
        vector<vector<int>> res;

        if(!root)
        {
            return res;
        }

        vector<int> head_val(1, root->val);
        res.push_back(head_val);
        vector<TreeNode*> curNodeList;
        curNodeList.push_back(root);

        while(curNodeList.size() > 0)
        {
            vector<TreeNode*> tempNodeList;
            vector<int> temp_res;
            
            for (int i = 0; i < curNodeList.size(); i++)
            {
                if (curNodeList[i]->left)
                {
                    tempNodeList.push_back(curNodeList[i]->left);
                    temp_res.push_back(curNodeList[i]->left->val);
                }
                if (curNodeList[i]->right)
                {
                    tempNodeList.push_back(curNodeList[i]->right);
                    temp_res.push_back(curNodeList[i]->right->val);
                }
            }

            curNodeList = tempNodeList;
            if(temp_res.size() > 0)
            {
                res.push_back(temp_res);
            }
        }
        return res;
    }
};

File: leetcode_hot_100/levelOrder.py
# hot100-lc 102. 层序遍历

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        ret = [[root.val]]
        curNodes = [root]
        while len(curNodes) > 0:
            tmp_ret = []
            nextLevel = []
            for node in curNodes:
                if node.left:
                    tmp_ret.append(node.left.val)
                    nextLevel.append(node.left)
                if node.right:
                    tmp_ret.append(node.right.val)
                    nextLevel.append(node.right)
            if len(tmp_ret) > 0:
                ret.append(tmp_ret)
            curNodes = nextLevel
        return ret
